lib/cloud_controller/deployment_updater/actions/scale.rb

```diff
@@ -36,36 +38,86 @@ def call
             end
 
             ScaleDownCanceledProcesses.new(deployment).call
-            ScaleDownOldProcess.new(deployment, oldest_web_process_with_instances, desired_old_instances).call
 
-            deploying_web_process.update(instances: desired_new_instances)
+            scale_down_old_processes if can_downscale?
+
+            if can_scale?
+              deploying_web_process.update(instances: desired_new_instances)
+              deployment.update(last_healthy_at: Time.now)
+            end
           end
         end
 
         private
 
-        def desired_old_instances
-          [(oldest_web_process_with_instances.instances - deployment.max_in_flight), 0].max
+        def scale_down_old_processes
+          instances_to_reduce = non_deploying_web_processes.map(&:instances).sum - desired_non_deploying_instances
+
+          return if instances_to_reduce <= 0
+
+          non_deploying_web_processes.each do |process|
+            if instances_to_reduce < process.instances
+              ScaleDownOldProcess.new(deployment, process, process.instances - instances_to_reduce).call
+              break
+            end
+
+            instances_to_reduce -= process.instances
+            ScaleDownOldProcess.new(deployment, process, 0).call
+          end
+        end
+
+        def can_scale?
+          starting_instances.count < deployment.max_in_flight && unhealthy_instances.count == 0
+        rescue CloudController::Errors::ApiError # the instances_reporter re-raises InstancesUnavailable as ApiError
+          logger.info("skipping-deployment-update-for-#{deployment.guid}")
+          false
+        end
+
+        def can_downscale?
+          non_deploying_web_processes.map(&:instances).sum > desired_non_deploying_instances
+        rescue CloudController::Errors::ApiError # the instances_reporter re-raises InstancesUnavailable as ApiError
+          logger.info("skipping-deployment-update-for-#{deployment.guid}")
+          false
+        end
+
+        def desired_non_deploying_instances
+          [deployment.original_web_process_instance_count - routable_instances.count, 0].max
         end
 
         def desired_new_instances
-          [deploying_web_process.instances + deployment.max_in_flight, deployment.original_web_process_instance_count].min
+          [deploying_web_process.instances + deployment.max_in_flight - starting_instances.count, deployment.original_web_process_instance_count].min
```

Comment:
**question/non-blocking** How come we need to minus the starting instance count here to determine the desired new instances? Shouldn't those instances not matter for the desired state?

Reply:
I think that would be the case if `deploying_web_process.instances` were the actual state, but it's the desired state. So we subtract starting instances to determine how many of the new max-in-flight instances are finished, and only increase by that amount (since they are no longer 'in flight')

Reply:
I changed this line to only look at routable_instances, but added a safeguard that prevents the scaling action from acting if there are fewer routable instances than there should be